### PR TITLE
Preloading story bookend

### LINKED ISSUE #12507: Bookend configuration is not pre-loaded
Currently loading until bookend needs to be opened. 

Should be preloaded ASAP, possibly on the same level as page preload based on distance.

### PR DESCRIPTION
**Pull request by gmajoulet**
Preloads the bookend when navigating to the last page.

- On desktop, the bookend is always loaded since it's required for the topbar sharing options
- On mobile, it is now preloaded
- New unit tests + re writing/enabling a few basic old ones (this needs more work, tracked in #11639)
- Skipping one test that fails and needs to be re written (the code still works)

Fixes #12507

### CODE REVIEW

File: extensions/amp-story/0.1/amp-story.js [hunk 1]
```diff
@@ -1092,6 +1100,18 @@ export class AmpStory extends AMP.BaseElement {
...
 
 
+  /**
+   * Preloads the bookend config if on the last page.
+   * @private
+   */
+  maybePreloadBookend_() {
+    const pageIndex = this.getPageIndex(this.activePage_);
+    if (pageIndex + 1 >= this.getPageCount()) {
+      this.loadBookendConfig_();
```
Comment (newmuis) [inline extensions/amp-story/0.1/amp-story.js]: I think we may as well call `this.buildBookend_()` here to actually pre-render it as well.

**Comment by choumx** [comment]
You can locally run `gulp lint` for lint errors and `gulp check-all` for all presubmit checks.

File: extensions/amp-story/0.1/amp-story.js (hunk 1 above)
Comment (gmajoulet) [reply extensions/amp-story/0.1/amp-story.js]: Done

**Comment by gmajoulet** [comment]
PTAL

File: extensions/amp-story/0.1/test/test-amp-story.js [hunk 2]
```diff
@@ -221,15 +305,6 @@ describes.realWin('amp-story', {
     expect(hideBookendStub).to.have.been.calledOnce;
   });
 
-  // TODO(newmuis/amphtml-story#187): Re-enable this test.
-  it.skip('should return a valid page count', () => {
```
Comment (newmuis) [inline extensions/amp-story/0.1/test/test-amp-story.js]: Any particular reason to remove this?

File: extensions/amp-story/0.1/test/test-amp-story.js [hunk 3]
```diff
@@ -241,18 +316,6 @@ describes.realWin('amp-story', {
     });
   });
 
-  // TODO(newmuis/amphtml-story#187): Re-enable this test.
-  it.skip('should return all pages', () => {
```
Comment (newmuis) [inline extensions/amp-story/0.1/test/test-amp-story.js]: Any particular reason to remove this?

File: extensions/amp-story/0.1/test/test-amp-story.js (hunk 2 above)
Comment (gmajoulet) [reply extensions/amp-story/0.1/test/test-amp-story.js]: I moved it to the first test. Should have kept the right order to help the review

File: extensions/amp-story/0.1/test/test-amp-story.js (hunk 3 above)
Comment (gmajoulet) [reply extensions/amp-story/0.1/test/test-amp-story.js]: This method no longer exists

File: extensions/amp-story/0.1/test/test-amp-story.js [hunk 4]
```diff
@@ -49,6 +50,7 @@ describes.realWin('amp-story', {
     return Array(count).fill(undefined).map((unused, i) => {
       const page = win.document.createElement('amp-story-page');
       page.id = opt_ids && opt_ids[i] ? opt_ids[i] : `-page-${i}`;
+      page.getImpl = () => Promise.resolve(new AmpStoryPage(page));
```
Comment (choumx) [inline extensions/amp-story/0.1/test/test-amp-story.js]: Is this still necessary with a file-level import (above)?

File: extensions/amp-story/0.1/test/test-amp-story.js [hunk 5]
```diff
@@ -13,12 +13,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import {AnalyticsTrigger} from '../analytics';
 import {AmpStory} from '../amp-story';
+import {AmpStoryPage} from '../amp-story-page';
```
Comment (choumx) [inline extensions/amp-story/0.1/test/test-amp-story.js]: This works without the file-level import? `import '../amp-story'`?

Comment (gmajoulet) [reply extensions/amp-story/0.1/test/test-amp-story.js]: Yes, the amp-story-page is properly instantiated

File: extensions/amp-story/0.1/test/test-amp-story.js (hunk 4 above)
Comment (gmajoulet) [reply extensions/amp-story/0.1/test/test-amp-story.js]: Unfortunately yes, without this line it'd break, even with the file-level import

Comment (choumx) [reply extensions/amp-story/0.1/test/test-amp-story.js]: Something's not right, but ok for now if it works.

**Pull request merged**